### PR TITLE
Add ZMQ IpfixReceiver subclass

### PR DESCRIPTION
**Pull request by bluca**
Add support to receive Ipfix/Netflow payloads over a ZMQ connection.
The support is optional and disabled by default. If enabled, adds
a build dependency on libczmq. CZMQ is a high-level API that builds
on libzmq, and exposes easy and nice to use APIs to communicate using
the ZeroMQ protocol.
For more information about ZeroMQ, please see:
http://zeromq.org/
https://github.com/zeromq/czmq

### CODE REVIEW

**Comment by bluca** [comment]
A bit more background on this:

The ZeroMQ project provides a set of libraries to use ZMTP, a protocol for setting up communication channels over TCP, UDP, IPC or shared memory, all using the same API. A variety of patterns and socket "types" are made available.

One of those patterns is the publisher-subscriber. This is what I implemented here. The ZMQ Ipfix Receiver can subscribe to one or more channels made available by one or  more publishers.
This pattern is good because it's resilient, offers low latency and high throughput, and doesn't require an handshake.

This makes it ideal for software dataplanes that generate reports and want to send them away quickly.

**Comment by bluca** [comment]
@ogasser - ping :-)

File: README [hunk 1]
```diff
@@ -28,6 +28,8 @@ The following packages are optional:
     ==> cmake option SUPPORT_MYSQL
  - libgsl-dev (for connection-based sampling with Bloom filters)
     ==> cmake option USE_GSL
+ - libczmq-dev (for receiving Ipfix reports over ZMQ)
```
Comment (ogasser) [inline README]: Please change Ipfix to IPFIX

File: CMakeLists.txt [hunk 2]
```diff
@@ -114,6 +116,27 @@ TARGET_LINK_LIBRARIES(vermont
...
+
+OPTION(SUPPORT_ZMQ "Enable ZMQ support" OFF)
+IF (SUPPORT_ZMQ)
+	FIND_PACKAGE(PkgConfig REQUIRED)
+	PKG_CHECK_MODULES(CZMQ libczmq)
+	IF (NOT CZMQ_FOUND)
+		MESSAGE(FATAL_ERROR "Could not find libczmq libraries.")
+	ENDIF (NOT CZMQ_FOUND)
+ENDIF (SUPPORT_ZMQ)
+IF (CZMQ_FOUND)
```
Comment (ogasser) [inline CMakeLists.txt]: Please add `MESSAGE(STATUS "Found libczmq libraries")`

File: README (hunk 1 above)
Comment (bluca) [inline README]: Done

File: CMakeLists.txt (hunk 2 above)
Comment (bluca) [inline CMakeLists.txt]: Done

File: src/modules/ipfix/CollectorCfg.h [hunk 3]
```diff
@@ -170,6 +192,10 @@ class CollectorCfg
 	uint16_t mtu;
 	uint32_t buffer;
 	std::set<std::string> peerFqdns;
+	std::vector<std::string> zmqEndpoints;
+	std::vector<std::string> zmqPubSubChannel;
```
Comment (ogasser) [inline src/modules/ipfix/CollectorCfg.h]: Could you change this to `zmqPubSubChannels` (plural) for consistency?

File: src/modules/ipfix/IpfixReceiverZmq.cpp [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+
+	// Stop CZMQ from hijacking signal handling
+	zsys_handler_set(NULL);
+
+	zpoller = zpoller_new(NULL);
+	if (!zpoller) {
+		THROWEXCEPTION("Could not create ZMQ poller, cannot start ZMQ Receiver");
+	}
+
+	for (std::vector<std::string>::iterator i = endpoints.begin();
```
Comment (ogasser) [inline src/modules/ipfix/IpfixReceiverZmq.cpp]: Can you join this and the following line, looks cleaner that way imho.

File: src/modules/ipfix/IpfixReceiverZmq.cpp [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+			i != endpoints.end(); ++i) {
```
Comment (ogasser) [inline src/modules/ipfix/IpfixReceiverZmq.cpp]: Please use `i++` in your code.

File: src/modules/ipfix/IpfixReceiverZmq.cpp [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+		// If no channel is passed down, listen on everything (empty string)
+		zsock_t *sock = zsock_new_sub((*i).c_str(), channels.empty() ? "" : NULL);
+		if (!sock) {
+			THROWEXCEPTION("Could not connect ZMQ socket, cannot start ZMQ"
```
Comment (ogasser) [inline src/modules/ipfix/IpfixReceiverZmq.cpp]: Join this and the following line.

File: src/modules/ipfix/IpfixReceiverZmq.cpp [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+					"Receiver");
+		}
+
+		zsock_set_sndhwm(sock, zmq_high_watermark);
+		zsock_set_rcvhwm(sock, zmq_high_watermark);
+
+		for (std::vector<std::string>::iterator j = channels.begin();
```
Comment (ogasser) [inline src/modules/ipfix/IpfixReceiverZmq.cpp]: Join this line with the following.

File: src/modules/ipfix/IpfixReceiverZmq.cpp [hunk 8]
```diff
@@ -0,0 +1,175 @@
...
+				j != channels.end(); ++j) {
+			zsock_set_subscribe(sock, (*j).c_str());
+		}
+
+		if (0 != zpoller_add(zpoller, sock)) {
```
Comment (ogasser) [inline src/modules/ipfix/IpfixReceiverZmq.cpp]: Can you change this to `if (!zpoller_add(zpoller, sock)) {` ?

File: src/modules/ipfix/CollectorCfg.h (hunk 3 above)
Comment (bluca) [inline src/modules/ipfix/CollectorCfg.h]: Done

File: src/modules/ipfix/IpfixReceiverZmq.cpp (hunk 4 above)
Comment (bluca) [inline src/modules/ipfix/IpfixReceiverZmq.cpp]: Done

File: src/modules/ipfix/IpfixReceiverZmq.cpp (hunk 5 above)
Comment (bluca) [inline src/modules/ipfix/IpfixReceiverZmq.cpp]: Done

File: src/modules/ipfix/IpfixReceiverZmq.cpp (hunk 6 above)
Comment (bluca) [inline src/modules/ipfix/IpfixReceiverZmq.cpp]: Done

File: src/modules/ipfix/IpfixReceiverZmq.cpp [hunk 9]
```diff
@@ -0,0 +1,171 @@
...
+
+/**
+ * Frees memory used by a IpfixReceiverZmq.
+ */
+IpfixReceiverZmq::~IpfixReceiverZmq()
+{
+	zpoller_destroy(&zpoller);
+
+	for (std::vector<zsock_t *>::iterator i = zmq_sockets.begin(); i != zmq_sockets.end(); i++) {
+		zsock_destroy(&(*i));
```
Comment (ogasser) [inline src/modules/ipfix/IpfixReceiverZmq.cpp]: Join this with the following line.

File: src/modules/ipfix/IpfixReceiverZmq.cpp (hunk 7 above)
Comment (bluca) [inline src/modules/ipfix/IpfixReceiverZmq.cpp]: Done

File: src/modules/ipfix/IpfixReceiverZmq.cpp [hunk 10]
```diff
@@ -0,0 +1,171 @@
...
+	SensorManager::getInstance().removeSensor(this);
+}
+
+/**
+ * ZMQ specific listener function. This function is called by @c listenerThread()
+ */
+void IpfixReceiverZmq::run()
+{
+	// create sourceId
+	boost::shared_ptr<IpfixRecord::SourceID> sourceID(
```
Comment (ogasser) [inline src/modules/ipfix/IpfixReceiverZmq.cpp]: Join lines.

File: src/modules/ipfix/IpfixReceiverZmq.cpp [hunk 11]
```diff
@@ -0,0 +1,171 @@
...
+		zframe_destroy(&current_frame);
+
+		//Frame 1..n: flow reports
+		size_t num_frames = zmsg_size(msg);
+		for (size_t j = 0; j < num_frames; j++) {
+			current_frame = zmsg_pop(msg);
+			boost::shared_array<uint8_t> data(new uint8_t[zframe_size(current_frame)]);
+			memcpy(data.get(), (uint8_t *)zframe_data(current_frame), zframe_size(current_frame));
+
+			++statReceivedPackets;
```
Comment (ogasser) [inline src/modules/ipfix/IpfixReceiverZmq.cpp]: Please use post-increment if there is no need to use pre.

Comment (bluca) [inline src/modules/ipfix/IpfixReceiverZmq.cpp]: Done

File: src/modules/ipfix/IpfixReceiverZmq.cpp (hunk 8 above)
Comment (bluca) [inline src/modules/ipfix/IpfixReceiverZmq.cpp]: Done (without the "!", as the success case returns 0)

File: src/modules/ipfix/IpfixReceiverZmq.cpp (hunk 10 above)
Comment (bluca) [inline src/modules/ipfix/IpfixReceiverZmq.cpp]: Done

File: src/modules/ipfix/IpfixReceiverZmq.cpp [hunk 12]
```diff
@@ -0,0 +1,170 @@
...
+		for (size_t j = 0; j < num_frames; j++) {
+			current_frame = zmsg_pop(msg);
+			boost::shared_array<uint8_t> data(new uint8_t[zframe_size(current_frame)]);
+			memcpy(data.get(), (uint8_t *)zframe_data(current_frame), zframe_size(current_frame));
+
+			statReceivedPackets++;
+
+			// send packet to all packet processors
+			mutex.lock();
+			for (std::list<IpfixPacketProcessor*>::iterator i =
```
Comment (ogasser) [inline src/modules/ipfix/IpfixReceiverZmq.cpp]: Merge lines

File: src/modules/ipfix/IpfixReceiverZmq.cpp [hunk 13]
```diff
@@ -0,0 +1,170 @@
+/*
+ * Vermont ZMQ observer module
```
Comment (ogasser) [inline src/modules/ipfix/IpfixReceiverZmq.cpp]: observer -> receiver

File: src/modules/ipfix/IpfixReceiverZmq.hpp [hunk 14]
```diff
@@ -0,0 +1,63 @@
+/*
+ * Vermont ZMQ observer module
```
Comment (ogasser) [inline src/modules/ipfix/IpfixReceiverZmq.hpp]: observer -> receiver

File: src/modules/ipfix/IpfixReceiverZmq.cpp (hunk 12 above)
Comment (bluca) [inline src/modules/ipfix/IpfixReceiverZmq.cpp]: Done

File: src/modules/ipfix/IpfixReceiverZmq.cpp (hunk 13 above)
Comment (bluca) [inline src/modules/ipfix/IpfixReceiverZmq.cpp]: Done

File: src/modules/ipfix/IpfixReceiverZmq.hpp (hunk 14 above)
Comment (bluca) [inline src/modules/ipfix/IpfixReceiverZmq.hpp]: Done

**Comment by bluca** [comment]
Thanks!

I'll make sure to follow the recommendations about code style for the next PRs.

**Pull request merged**